include/rmm/mr/device/failure_alternate_resource_adaptor.hpp

```diff
@@ -0,0 +1,159 @@
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cstddef>
+#include <mutex>
+#include <unordered_set>
+
+namespace RMM_NAMESPACE {
+namespace mr {
+/**
+ * @addtogroup device_resource_adaptors
+ * @{
+ * @file
+ */
+
+/**
+ * @brief A device memory resource that use an alternate upstream resource when the primary throw a
+ * specified exception type.
+ *
+ * An instance of this resource must be constructed with two existing upstream resource to
+ * satisfy allocation requests.
+ *
+ * @tparam ExceptionType The type of exception that this adaptor should respond to.
+ */
+template <typename ExceptionType = rmm::out_of_memory>
+class failure_alternate_resource_adaptor final : public device_memory_resource {
+ public:
+  using exception_type = ExceptionType;  ///< The type of exception this object catches/throws
+
+  /**
+   * @brief Construct a new `failure_alternate_resource_adaptor` using `primary_upstream` as the
+   * primary resource to satisfy allocation requests and if that fails, use `alternate_upstream`
+   * as an alternate
+   *
+   * @param primary_upstream The primary resource used for allocating/deallocating device memory
+   * @param alternate_upstream The alternate resource used for allocating/deallocating device memory
+   * memory
+   */
+  failure_alternate_resource_adaptor(device_async_resource_ref primary_upstream,
+                                     device_async_resource_ref alternate_upstream)
+    : primary_upstream_{primary_upstream}, alternate_upstream_{alternate_upstream}
+  {
+  }
+
+  failure_alternate_resource_adaptor()                                          = delete;
+  ~failure_alternate_resource_adaptor() override                                = default;
+  failure_alternate_resource_adaptor(failure_alternate_resource_adaptor const&) = delete;
+  failure_alternate_resource_adaptor& operator=(failure_alternate_resource_adaptor const&) = delete;
+  failure_alternate_resource_adaptor(failure_alternate_resource_adaptor&&) noexcept =
+    default;  ///< @default_move_constructor
+  failure_alternate_resource_adaptor& operator=(failure_alternate_resource_adaptor&&) noexcept =
+    default;  ///< @default_move_assignment{failure_alternate_resource_adaptor}
+
+  /**
+   * @briefreturn{rmm::device_async_resource_ref to the upstream resource}
+   */
+  [[nodiscard]] rmm::device_async_resource_ref get_upstream_resource() const noexcept
+  {
+    return primary_upstream_;
+  }
+
+  /**
+   * @briefreturn{rmm::device_async_resource_ref to the alternate upstream resource}
+   */
+  [[nodiscard]] rmm::device_async_resource_ref get_alternate_upstream_resource() const noexcept
+  {
+    return alternate_upstream_;
+  }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource.
+   *
+   * @throws `exception_type` if the requested allocation could not be fulfilled
+   * by the primary or the alternate upstream resource.
```

Comment:
I think it's `ExceptionType`. But actually I don't think you can say what type of exception will be thrown by `alternate_upstream` if it fails to allocate. It could be a CUDA error, or it could be `rmm::out_of_memory`, or some other exception.

The alternate upstream should document what exceptions it may throw.

Reply:
fixed